components/scream/src/share/io/scorpio_input.cpp

```diff
@@ -0,0 +1,277 @@
+  for (auto& tag_ii : vec_of_dims)
+  {
+    io_decomp_tag += "-" + tag_ii;
+  }
+  return io_decomp_tag;
+}
+/* ---------------------------------------------------------- */
+void AtmosphereInput::register_views()
+{
+  using namespace scream;
+  using namespace scream::scorpio;
+
+  // Cycle through all fields and register.
+  for (auto const& name : m_fields)
+  {
+    auto fmap = m_field_repo->get_field(name, m_grid_name);
+    // If the "averaging type" is instant then just need a ptr to the view.
+    auto view_d = fmap.get_view();
+    // Create a local copy of view to be stored by input stream.
+    auto view_copy_h = Kokkos::create_mirror_view( view_d );
+    view_type view_copy("",view_copy_h.extent(0));
+    Kokkos::deep_copy(view_copy, view_d);
+    m_view_local.emplace(name,view_copy);
+  }
+  if (m_is_rhist)
+  {
+    view_type l_view("",1);
+    m_view_local.emplace("avg_count",l_view);
+  }
+}
+/* ---------------------------------------------------------- */
+void AtmosphereInput::set_degrees_of_freedom()
+{
+/* 
+ * Use information from the grids manager to determine which indices for each field are associated
+ * with this computational core.
+ */
+  using namespace scream;
+  using namespace scream::scorpio;
+  // Cycle through all fields and set dof.
+  for (auto const& name : m_fields)
+  {
+    auto fmap = m_field_repo->get_field(name, m_grid_name);
+    auto& fid  = fmap.get_header().get_identifier();
+    // Given dof_len and n_dim_len it should be possible to create an integer array of "global input indices" for this
+    // field and this rank. For every column (i.e. gid) the PIO indices would be (gid * n_dim_len),...,( (gid+1)*n_dim_len - 1).
+    std::vector<Int> var_dof = get_var_dof(fid.get_layout().size(), fid.get_layout().has_tag(FieldTag::Column));
+    set_dof(m_filename,name,var_dof.size(),var_dof.data());
+    m_dofs.emplace(std::make_pair(name,var_dof.size()));
+  }
+  if (m_is_rhist)
+  {
+    Int var_dof[1] = {0};
+    set_dof(m_filename,"avg_count",1,var_dof);
+    m_dofs.emplace(std::make_pair("avg_count",1));
+  }
+  /* TODO: 
+   * Gather DOF info directly from grid manager
+  */
+} // set_degrees_of_freedom
+/* ---------------------------------------------------------- */
+std::vector<Int> AtmosphereInput::get_var_dof(const int dof_len, const bool has_cols)
+{
+  std::vector<Int> var_dof;
+  int num_cols;
+  // Gather the column degrees of freedom for this variable
+  // Total number of values represented by this rank for this field is given by the dof_len.
+  // For a SCREAM Physics grid, only the total number of columns is decomposed over MPI ranks.  The global id (gid)
+  // is stored here as gids_hst. Thus, for this field, the total number of dof's in the other dimensions (i.e. levels)
+  // can be found by taking the quotient of dof_len and the length of gids_hst.
+  auto gids_dev = m_gm->get_grid(m_grid_name)->get_dofs_gids();
+  auto gids_hst = Kokkos::create_mirror_view( gids_dev );
```

Comment:
You might consider storing gids_hst in your class, so you don't do a deep copy every time you pull a variable.

And btw, it took me a minute to realize `hst` doesn't stand for "history" (which in the I/O context is also a thing). Since saving one char is not much, do you mind using `_host`  for host variables?

Reply:
Yes, good point.  I was trying to keep the names the same length for aesthetics :-).  I've changed to "host"